docs/versioning.md

```diff
@@ -0,0 +1,463 @@
+# Versioning
+
+This section describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+## AWS S3 Bucket Versioning
+
+An object in a versioning enabled bucket can have multiple versions, whereas an
+object can have only a single version in a standard bucket without versioning.
+
+With the standard S3 API, any update (PUT, DELETE) to an object in a versioning
+enabled bucket would create a new version of that object. Read operations such
+as GET and LIST would see the latest version of the objects. The table below
+shows such example, a `PUT foo` generates a new version `v2` and keeps `v1`;
+GET and LIST return the latest version of `foo`, which is `v2` in this case.
+
+| key | version | value |
+|-----|:-------:|:-----:|
+| foo |   v2    |   B   |
+| foo |   v1    |   A   |
+
+To work with all existing versions of an object, S3 provides an extended
+API which uses the same operations of the standard API, but with versioning
+information specified. For example, in order to get the version `v1` of `foo`,
+users have to specify that version in the query, such as
+`GET /foo?versionId=v1`.
+
+A bucket can be non-versioning or versioning. Any objects placed in a bucket
+before versioning is enabled are considered to be the "null" version. When
+versioning is enabled, the "null" version will be kept and continue to be
+identified as the "null" version. However, if versioning is suspended, the
+"null" version may be overwritten with a new "null" version. All versions with
+a version ID other than "null" are retained when versioning is suspended, even
+when there are new puts on the object.
+
+## Implementation of Bucket Versioning in S3
+
+### Roles of Metadata Engine and API Component
+
+Each version of an object is stored as a separate key in metadata. The API
+component of S3 manages the put and delete of version data, sends instructions
+to create, retrieve, overwrite or delete a version's metadata to the metadata
+backend, and tracks extra information about a version, such as whether it is a
+delete marker or a null version.
+
+The metadata backend is responsible for handling the create, overwrite and
+delete operations of version metadata upon receiving instructions. We use
+version identifiers as the suffix for the keys of the object versions, and a
+special version to represent the latest version. An example of what these
+keys may look like in metadata, for an object `foo/bar` with three versions:
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar\098506163554375999999PARIS  0.a430a1f85c6ec |
+| foo/bar\098506163554373999999PARIS  0.41b510cd0fdf8 |
+| foo/bar\098506163554373999998PARIS  0.f9b82c166f695 |
```

Comment:
The key `foo/bar` is a special version because it's not followed by the version id suffix, but I agree it's not very clear for readers at this point. I added a note to make it clearer and alluded to where it's described in more detail in the section "Master Versions"